File: extension/core/src/main/java/io/holunda/camunda/bpm/data/adapter/basic/ReadWriteAdapterVariableMap.java
package io.holunda.camunda.bpm.data.adapter.basic;

import java.util.Optional;
import org.camunda.bpm.engine.variable.VariableMap;

/**
 * Read-write adapter for variable map.
 *
 * @param <T> type of value.
 */
public class ReadWriteAdapterVariableMap<T> extends AbstractBasicReadWriteAdapter<T> {

  private final VariableMap variableMap;

  /**
   * Constructs the adapter.
   *
   * @param variableMap variable map to access.
   * @param variableName variable to access.
   * @param clazz class of variable value.
   */
  public ReadWriteAdapterVariableMap(VariableMap variableMap, String variableName, Class<T> clazz) {
    super(variableName, clazz);
    this.variableMap = variableMap;
  }

  @Override
  public Optional<T> getOptional() {
    return Optional.ofNullable(getOrNull(variableMap.get(variableName)));
  }

  @Override
  public void set(T value, boolean isTransient) {
    variableMap.putValueTyped(variableName, getTypedValue(value, isTransient));
  }

  @Override
  public Optional<T> getLocalOptional() {
    throw new UnsupportedOperationException("Can't get a local variable on a variable map");
  }

  @Override
  public void setLocal(T value, boolean isTransient) {
    throw new UnsupportedOperationException("Can't set a local variable on a variable map");
  }

  @Override
  public void remove() {
    variableMap.remove(variableName);
  }

  @Override
  public void removeLocal() {
    throw new UnsupportedOperationException("Can't set a local variable on a variable map");
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/basic/ReadWriteAdapterVariableMap.kt
package io.holunda.camunda.bpm.data.adapter.basic

import org.camunda.bpm.engine.variable.VariableMap
import java.util.*

/**
 * Read-write adapter for variable map.
 *
 * @param [T] type of value.
 * @param variableMap  variable map to access.
 * @param variableName variable to access.
 * @param clazz        class of variable value.
 */
class ReadWriteAdapterVariableMap<T : Any?>(private val variableMap: VariableMap, variableName: String, clazz: Class<T>) :
  AbstractBasicReadWriteAdapter<T>(variableName, clazz) {

  override fun getOptional(): Optional<T> {
    @Suppress("UNCHECKED_CAST")
    return Optional.ofNullable(getOrNull(variableMap[variableName])) as Optional<T>
  }

  override fun set(value: T, isTransient: Boolean) {
    variableMap.putValueTyped(variableName, getTypedValue(value, isTransient))
  }

  override fun getLocalOptional(): Optional<T> {
    throw UnsupportedOperationException("Can't get a local variable on a variable map")
  }

  override fun setLocal(value: T, isTransient: Boolean) {
    throw UnsupportedOperationException("Can't set a local variable on a variable map")
  }

  override fun remove() {
    variableMap.remove(variableName)
  }

  override fun removeLocal() {
    throw UnsupportedOperationException("Can't set a local variable on a variable map")
  }
}
